src/apps/external_executor/external_executor.cpp

```diff
@@ -207,47 +227,98 @@ namespace externalexecutor
       auto executor_auth_policy = std::make_shared<ExecutorAuthPolicy>();
       ccf::AuthnPolicies executor_only{executor_auth_policy};
 
-      auto start = [this](
-                     ccf::endpoints::CommandEndpointContext& ctx,
-                     google::protobuf::Empty&& payload)
-        -> ccf::grpc::GrpcAdapterResponse<
-          externalexecutor::protobuf::OptionalRequestDescription> {
+      auto activate =
+        [this](
+          ccf::endpoints::CommandEndpointContext& ctx,
+          google::protobuf::Empty&& payload,
+          ccf::grpc::StreamPtr<externalexecutor::protobuf::Work>&& out_stream)
+        -> ccf::grpc::GrpcAdapterStreamingResponse {
+        const auto executor_id = get_caller_executor_id(ctx);
+        const auto it = active_executors.find(executor_id);
+        if (it != active_executors.end())
+        {
+          return ccf::grpc::make_error(
+            GRPC_STATUS_FAILED_PRECONDITION,
+            fmt::format(
+              "Executor {} is already active, cannot Activate again",
+              executor_id));
+        }
+        else
+        {
+          // Signal to this executor that its activation has succeeded
+          externalexecutor::protobuf::Work work;
+          work.mutable_activated();
+          out_stream->stream_msg(work);
+
+          active_executors.emplace_hint(
+            it,
+            executor_id,
+            ExecutorInfo{
+              {},
+              ccf::grpc::detach_stream(
+                ctx.rpc_ctx, std::move(out_stream), [this, executor_id]() {
+                  auto search = active_executors.find(executor_id);
```

Comment:
Should there be a lock to access `active_executors` here and elsewhere?

Reply:
Somewhere yes, but I'm planning to replace this container structure shortly as well.